packages/datadog-plugin-aws-sdk/src/base.js

```diff
@@ -0,0 +1,91 @@
+'use strict'
+
+const Tags = require('opentracing').Tags
+const analyticsSampler = require('../../dd-trace/src/analytics_sampler')
+const awsHelpers = require('./aws_helpers')
+
+function createWrapRequest (tracer, config) {
+  config = normalizeConfig(config)
+
+  return function wrapRequest (request) {
+    return function requestWithTrace (operation, params, cb) {
+      const serviceName = awsHelpers.normalizeServiceName(this)
+
+      const baseTags = {
+        [Tags.SPAN_KIND]: 'client',
+        'span.type': 'http',
+        'service.name': config.service || `${tracer._service}-aws`,
+        'aws.agent': 'js-aws-sdk',
+        'aws.operation': operation,
+        'aws.region': (request.httpRequest && request.httpRequest.region) || this.config.region,
+        'aws.service': serviceName,
+        'component': serviceName
+      }
+
+      const childOf = tracer.scope().active()
+      const span = tracer.startSpan('aws.http', {
+        childOf,
+        tags: baseTags
+      })
+
+      awsHelpers.addAwsRequestTags(span, request, operation, params, serviceName)
+
+      analyticsSampler.sample(span, config.analytics)
+
+      // https://github.com/awsdocs/aws-javascript-developer-guide-v2/blob/
+      // master/doc_source/using-a-callback-function.md#using-an-anonymous-callback-function
+      if (typeof cb === 'function') {
+        return tracer.scope().activate(span, () => {
+          return request.call(this, operation, params, awsHelpers.wrapCallback(tracer, span, cb, childOf, config))
+        })
+      } else {
+        const awsReq = request.apply(this, arguments)
+        const boundAwsReq = tracer.scope().bind(awsReq, span)
+
+        boundAwsReq.on('send', response => {
+          tracer.scope().activate(span)
+        })
+
+        // https://github.com/awsdocs/aws-javascript-developer-guide-v2/blob/
+        // master/doc_source/using-a-response-event-handler.md#the-complete-event
+        boundAwsReq.on('complete', response => {
+          // response is same AWS.Response object
+          // as https://github.com/aws/aws-sdk-js/issues/781#issuecomment-156250427
+          awsHelpers.addAwsResponseTags(span, response)
+          config.hooks.addCustomTags(span, params)
+          awsHelpers.finish(span, response.error, config)
+        })
+
+        return boundAwsReq
+      }
+    }
+  }
+}
+
+function normalizeConfig (config) {
+  const hooks = getHooks(config)
+
+  return Object.assign({}, config, {
+    hooks
+  })
+}
+
+function getHooks (config) {
+  const noop = () => {}
+  const addCustomTags = (config.hooks && config.hooks.addCustomTags) || noop
+
+  return { addCustomTags }
+}
+
+module.exports = [
+  {
+    name: 'aws-sdk',
+    versions: ['>=2.3.0'],
```

Comment:
So 2.3.0 introduced the .promise() function which is the suggested way to use the package now, this was in `03/2016`.

  - https://aws.amazon.com/releasenotes/release-aws-sdk-for-javascript-v2-3-0/?tag=releasenotes%23keywords%23javascript
  - https://github.com/aws/aws-sdk-js/releases/tag/v2.3.0

And additionally, I think the main usage of this would be in Lambda's like here:

  - https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html

Which use `2.585.0` for both Node.js 10 and 12

However, for `>=2.0` I think mainly I would just need to do use `util.promise` instead of `promise()` for scaffolding in the before/after Each hook of the tests and then only test the `.promise()` with specific versions. Not a huge lift but just will need to implement, i don't think anything changes regarding instrumentation. I can get started on that.

For 1.x support, I think there will be more significant changes required as there were a number of breaking changes, and I'm not really sure whether many ppl are using 1.x? I haven't scoped out the changes but I'd have to work through this document here 
  - https://github.com/aws/aws-sdk-js/blob/master/UPGRADING.md 

Let me know if we feel like this is a requirement here and I can start to scope this out.

Lastly it's worth pointing out that 3.x is being worked on and is coming at some point so that is probably something to keep an eye on as it looks to be a [complete rewrite in typescript](https://github.com/aws/aws-sdk-js-v3). We may want to restrict to >2, <3 in the meantime?

Reply:
updated to support >2.1, which ensures support before `.promise()` was introduced. didn't go all the way down to >2 because there were some bugs in 2.0.3 and earlier which make testing and general usage of the library very difficult, such as https://github.com/aws/aws-sdk-js/commit/a6c03caab714f4ddc2f6f731692d2c2ccc81444c